src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -806,11 +815,19 @@ def create_x509_certificate(self, builder, private_key, algorithm):
             isinstance(private_key, rsa.RSAPrivateKey)
         ):
             raise ValueError(
-                "MD5 is not a supported hash algorithm for EC/DSA certificates"
+                "MD5 is only (reluctantly) supported for RSA certificates"
             )
 
         # Resolve the signature algorithm.
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
+
+            # OpenSSL requires us to pass NULL for EVP_MD at this time, but
+            # will do ed25519ph (SHA512 hashing)
+            evp_md = self._ffi.NULL
+        else:
+            evp_md = self._evp_md_non_null_from_algorithm(algorithm)
```

Comment:
Ok, there is now officially too much duplicated code, can we consolidate the common portions of this into a function?